packages/order-utils/src/eip712_utils.ts

```diff
@@ -0,0 +1,95 @@
+     */
+    compileSchema(schema: EIP712Schema): Buffer {
+        const eip712Schema = EIP712Utils._encodeType(schema);
+        const eip712SchemaHashBuffer = crypto.solSHA3([eip712Schema]);
+        return eip712SchemaHashBuffer;
+    },
+    /**
+     * Merges the EIP712 hash of a struct with the DomainSeparator for 0x v2.
+     * @param   hashStruct the EIP712 hash of a struct
+     * @param   contractAddress the exchange contract address
+     * @return  The hash of an EIP712 message with domain separator prefixed
+     */
+    createEIP712Message(hashStruct: Buffer, contractAddress: string): Buffer {
+        const domainSeparatorHashBuffer = EIP712Utils._getDomainSeparatorHashBuffer(contractAddress);
+        const messageBuff = crypto.solSHA3([EIP191_PREFIX, domainSeparatorHashBuffer, hashStruct]);
+        return messageBuff;
+    },
+    pad32Address(address: string): Buffer {
+        const addressBuffer = ethUtil.toBuffer(address);
+        const addressPadded = EIP712Utils.pad32Buffer(addressBuffer);
+        return addressPadded;
+    },
+    pad32Buffer(buffer: Buffer): Buffer {
+        const bufferPadded = ethUtil.setLengthLeft(buffer, EIP712_VALUE_LENGTH);
+        return bufferPadded;
+    },
+    _getDomainSeparatorSchemaBuffer(): Buffer {
+        return EIP712Utils.compileSchema(EIP712_DOMAIN_SCHEMA);
+    },
+    _getDomainSeparatorHashBuffer(exchangeAddress: string): Buffer {
+        const domainSeparatorSchemaBuffer = EIP712Utils._getDomainSeparatorSchemaBuffer();
+        const encodedData = EIP712Utils._encodeData(EIP712_DOMAIN_SCHEMA, {
+            name: EIP712_DOMAIN_NAME,
+            version: EIP712_DOMAIN_VERSION,
+            verifyingContract: exchangeAddress,
+        });
+        const domainSeparatorHashBuff2 = crypto.solSHA3([domainSeparatorSchemaBuffer, ...encodedData]);
+        return domainSeparatorHashBuff2;
+    },
+    _encodeType(schema: EIP712Schema): string {
+        const namedTypes = _.map(schema.parameters, ({ name, type }) => `${type} ${name}`);
+        const namedTypesJoined = namedTypes.join(',');
+        const encodedType = `${schema.name}(${namedTypesJoined})`;
+        return encodedType;
+    },
+    _encodeData(schema: EIP712Schema, data: { [key: string]: any }): any {
+        const encodedValues = [];
+        for (const parameter of schema.parameters) {
+            const value = data[parameter.name];
+            if (parameter.type === EIP712Types.String || parameter.type === EIP712Types.Bytes) {
+                encodedValues.push(crypto.solSHA3([ethUtil.toBuffer(value)]));
+            } else if (parameter.type === EIP712Types.Uint256) {
+                encodedValues.push(value);
+            } else if (parameter.type === EIP712Types.Address) {
+                encodedValues.push(EIP712Utils.pad32Address(value));
+            } else {
+                throw new Error(`Unable to encode ${parameter.type}`);
+            }
+        }
+        return encodedValues;
+    },
```

Comment:
There is also `rawEncode` from `ethereumjs-abi` that can be used here, like in this example https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.js#L82. Advantage of using `rawEncode` is that it automatically handles all the different number types from Solidity correctly.

Reply:
This doesn't seem to like our BigNumbers. Can this be a future TODO or is it blocking the merge?

Reply:
This can for sure be a TODO.

I noticed the BigNumber get's confused when there are multiple instances of the library. It is very hard / impossible to synchronize those.